PowershellTools.Common/IntelliSense/CommandCompletionHelper.cs

```diff
@@ -0,0 +1,70 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+using System.Management.Automation.Language;
+using System.Management.Automation.Runspaces;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.Text;
+
+namespace PowerShellTools.Common.IntelliSense
+{
+    internal static class CommandCompletionHelper
+    {
+        public static CommandCompletion GetCommandCompletionList(string script, int caretPosition, Runspace runspace)
+        {
+            Ast ast;
+            Token[] tokens;
+            IScriptPosition cursorPosition;
+            GetCommandCompletionParameters(script, caretPosition, out ast, out tokens, out cursorPosition);
+            if (ast == null)
+            {
+                return null;
+            }
+
+            var ps = PowerShell.Create();
+            ps.Runspace = runspace;
+
+            return CommandCompletion.CompleteInput(ast, tokens, cursorPosition, null, ps);
+        }
+
+        public static void GetCommandCompletionParameters(string script, int caretPosition, out Ast ast, out Token[] tokens, out IScriptPosition cursorPosition)
+        {       
+            ParseError[] array;            
+            ast = Tokenize(script, out tokens, out array);            
+            if (ast != null)
+            {
+                //HACK: Clone with a new offset using private method... 

```

Comment:
Why do we need this hack? If it's a limitation on the PS/underlying APIs, perhaps we can ask them to change this for vNext?

If we need to access a private method, we have no guarantee it will have the same functionality, or even remain in future versions.  If this is unable to be worked-around, let's be sure we have a good chunk of tests around this.